app/aboot/fs_boot.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <debug.h>
+#include <target.h>
+#include <string.h>
+
+#include <lib/bio.h>
+#include <lib/fs.h>
+
+#include "fs_boot.h"
+
+static const char *bootable_parts[] = {
+	"system",
+	"cache",
+	"userdata",
+};
+static const int BOOTABLE_PART_COUNT = 3;
+
+static bool fsboot_bootable_part(char *label)
+{
+	int i;
+	for (i = 0; i < BOOTABLE_PART_COUNT; ++i)
+		if (strcmp(label, bootable_parts[i]) == 0)
+			return true;
+
+	return false;
+}
+
+/* Only iterate through files if target is NULL */
+static int fsboot_fs_load_img(char *dev_name, void* target, size_t sz)
+{
+	struct dirhandle *dirh;
+	struct dirent dirent;
+
+	char image_path[128] = "/mnt/";
+	int ret = 0, error = -2;
+
+	if (fs_mount("/mnt", "ext2", dev_name) >= 0) {
+		error = 0;
+		ret = fs_open_dir("/mnt", &dirh);
+		if (ret >= 0) {
+			while (fs_read_dir(dirh, &dirent) >= 0) {
+				if (!target) {
+					dprintf(SPEW, "| /%s/%s\n", dev_name, dirent.name);
+				} else if (strncmp(dirent.name, "boot.img", 7) == 0) {
+					strcpy(image_path, "/mnt/");
+					strcat(image_path, dirent.name);
+					dprintf(INFO, "Found boot image: %s : %s\n", dev_name, image_path);
+					fs_close_dir(dirh);
+					return fs_load_file(image_path, target, sz);
+				}
+			}
+			fs_close_dir(dirh);
+		} else {
+			if (!target)
+				dprintf(SPEW, "    fs_open_dir ret = %d\n", ret);
+			else
+				return ret;
+		}
+		fs_unmount("/mnt");
+	}
+	return -1;
+}
+
+static int fsboot_find_and_boot(int bdev_id, void* target, size_t sz)
+{
+	int i = 0, j = 0, ret = 0;
+	char dev_name[128];
+	bdev_t *dev = NULL;
+
+	/* HACK: There is no hd1p0 for some reason */
+	if (bdev_id == 1)
+		i = 1;
+
+	sprintf(dev_name, "hd%d", bdev_id);
+	dev = bio_open(dev_name);
+	if (!dev) {
+		dprintf(CRITICAL, "fs-boot: Can't open %s\n", dev_name);
+		return -1;
+	}
+	bio_close(dev);
```

Comment:
I guess it's fine, but would it make a difference if you skip opening the root and just try to open the partitions directly below?

Reply:
I want that to make a clear difference between those instead of printing an "empty" list in case a device is missing:
```
[620] fs-boot: Scanned devices:
[630] fs-boot: Can't open hd2   # <---------
[630] fs-boot: Looking at hd1:  # <---------
[630]   hd1p25:      system (  1536 MiB): 
[640]   hd1p29:       cache (   150 MiB): 
[640]   hd1p31:    userdata ( 12679 MiB): 
```